### PR TITLE
Cortex anit cleanup and module bootstraping

### PR DESCRIPTION
**Pull request by vEpiphyte**
* Reorganize how the Cortex startup works to pre-initialize all static data/variables as early as possible, pre-initalize modules prior to bringing storage online, and actually remove modules from the module list that fail their init or pre functions.  This also allows hooking in additional layer ctors.
* Update cormeodule unit tests.

### CODE REVIEW

File: synapse/lib/module.py [hunk 1]
```diff
@@ -103,18 +102,38 @@ def getModPath(self, *paths):
         dirn = self.getModDir()
         return s_common.genpath(dirn, *paths)
 
+    async def preCoreModule(self):
+        '''
+        Module implementers may over-ride this method to execute code
```
Comment (jnwatson) [inline synapse/lib/module.py]: over-ride -> override

File: synapse/tests/test_lib_module.py [hunk 2]
```diff
@@ -1,73 +1,107 @@
...
+            # modpaths are dynamically made on demand
+            self.false(os.path.isdir(foomod._modpath))
+            mpath = foomod.getModPath()
+            self.isin(os.path.join('mods', 'foo'), mpath)
+            self.true(os.path.isdir(foomod._modpath))
+
+            # preload a config file for the BarModule
+            dirn = s_common.gendir(core.dirn, 'mods', 'barmod')
+            s_common.yamlsave({'test': 1, 'duck': 'quack'}, dirn, 'conf.yaml')
+            # barmodule laods a layerctor
```
Comment (jnwatson) [inline synapse/tests/test_lib_module.py]: laods -> loads

File: synapse/cortex.py [hunk 3]
```diff
@@ -1719,6 +1762,15 @@ def getCoreInfo(self):
 
             mods.append(modu)
 
+            try:  # pragma: no cover
```
Comment (vEpiphyte) [inline synapse/cortex.py]: extra pragma - remove it

File: synapse/cortex.py [hunk 4]
```diff
@@ -1727,16 +1779,15 @@ def getCoreInfo(self):
 
     async def _initCoreMods(self):
 
-        for name, modu in self.modules.items():
+        for ctor, modu in list(self.modules.items()):
 
             try:
                 await s_coro.ornot(modu.initCoreModule)
-
             except asyncio.CancelledError:
```
Comment (vEpiphyte) [inline synapse/cortex.py]: add pragma

**Pull request merged**